cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,331 @@
+ */
+FlowLayout::~FlowLayout()
+{
+    QLayoutItem *item;
+    while ((item = FlowLayout::takeAt(0))) {
+        delete item;
+    }
+}
+
+/**
+ * @brief Indicates the layout's support for expansion in both horizontal and vertical directions.
+ * @return The orientations (Qt::Horizontal | Qt::Vertical) this layout can expand to fill.
+ */
+Qt::Orientations FlowLayout::expandingDirections() const
+{
+    return Qt::Horizontal | Qt::Vertical;
+}
+
+/**
+ * @brief Indicates that this layout's height depends on its width.
+ * @return True, as the layout adjusts its height to fit the specified width.
+ */
+bool FlowLayout::hasHeightForWidth() const
+{
+    return true;
+}
+
+/**
+ * @brief Calculates the required height to display all items within the specified width.
+ * @param width The available width for arranging items.
+ * @return The total height needed to fit all items in rows constrained by the specified width.
+ */
+int FlowLayout::heightForWidth(const int width) const
+{
+    int height = 0;
+    int rowWidth = 0;
+    int rowHeight = 0;
+
+    for (const QLayoutItem *item : items) {
+        if (item == nullptr || item->isEmpty()) {
+            continue;
+        }
+
+        int itemWidth = item->sizeHint().width() + horizontalSpacing();
+        if (rowWidth + itemWidth > width) { // Start a new row if the row width exceeds available width
+            height += rowHeight + verticalSpacing();
+            rowWidth = itemWidth;
+            rowHeight = item->sizeHint().height() + verticalSpacing();
+        } else {
+            rowWidth += itemWidth;
+            rowHeight = qMax(rowHeight, item->sizeHint().height());
+        }
+    }
+    height += rowHeight; // Add the final row's height
+    return height;
+}
+
+/**
+ * @brief Arranges layout items in rows within the specified rectangle bounds.
+ * @param rect The area within which to position layout items.
+ */
+void FlowLayout::setGeometry(const QRect &rect)
+{
+    QLayout::setGeometry(rect); // Sets the geometry of the layout based on the given rectangle.
+
+    int left, top, right, bottom;
+    getContentsMargins(&left, &top, &right, &bottom); // Retrieves the layout's content margins.
+
+    // Adjust the rectangle to exclude margins.
+    const QRect adjustedRect = rect.adjusted(+left, +top, -right, -bottom);
+
+    // Calculate the available width for items, considering either the adjusted rectangle's width
+    // or the parent scroll area width, if applicable.
+    const int availableWidth = qMax(adjustedRect.width(), getParentScrollAreaWidth());
+
+    // Arrange all rows of items within the available width and get the total height used.
+    const int totalHeight = layoutAllRows(adjustedRect.x(), adjustedRect.y(), availableWidth);
+
+    // If the layout's parent is a QWidget, update its minimum size to ensure it can accommodate
+    // the arranged items' dimensions.
+    if (QWidget *parentWidgetPtr = parentWidget()) {
```

Comment:
Is this a new shorthand annotation to set a value and see if it's non-null?

Reply:
Apparently. CLion suggested this as a quick-fix ("Variable assignment can be moved to if-statement")